lib/error-matcher.js

```diff
@@ -68,42 +68,61 @@ export default class ErrorMatcher extends EventEmitter {
 
   _parse() {
     this.currentMatch = [];
-    const self = this;
-    const matchFunction = function (match, i, string, regex) {
-      match.id = 'error-match-' + self.regex.indexOf(regex) + '-' + i;
-      this.push(match);
-    };
-    this.regex.forEach((regex) => {
-      if (typeof regex === 'function') {
-        regex(this.output).forEach(function (match) {
-          this.push(match);
-        }.bind(this.currentMatch));
-      } else {
-        require('xregexp').forEach(this.output, regex, matchFunction.bind(this.currentMatch));
-      }
+
+    // first run all functional matches
+    this.functions && this.functions.forEach((f, functionIndex) => {
+      this.currentMatch = this.currentMatch.concat(f(this.output).map((match, matchIndex) => {
+        match.id = 'error-match-function-' + functionIndex + '-' + matchIndex;
+        match.type = match.type || 'Error';
+        return match;
+      }));
+    });
+    // then for all match kinds
+    Object.keys(this.regex).forEach(kind => {
+      // run all matches
+      this.regex[kind] && this.regex[kind].forEach((regex, i) => {
+        regex && require('xregexp').forEach(this.output, regex, (match, matchIndex) => {
+          match.id = 'error-match-' + i + '-' + matchIndex;
+          match.type = kind;
+          this.currentMatch.push(match);
+        });
+      });
     });
 
     this.currentMatch.sort((a, b) => a.index - b.index);
 
     this.firstMatchId = (this.currentMatch.length > 0) ? this.currentMatch[0].id : null;
   }
 
-  set(regex, cwd, output) {
+  _prepareRegex(regex) {
     regex = regex || [];
     regex = (regex instanceof Array) ? regex : [ regex ];
 
-    this.regex = regex.map((r) => {
-      if (typeof r === 'function') {
-        return r;
-      }
+    return regex.map(r => {
       try {
         const XRegExp = require('xregexp');
         return XRegExp(r);
       } catch (err) {
         this.emit('error', 'Error parsing regex. ' + err.message);
         return null;
       }
-    }).filter(Boolean);
+    });
+  }
+
+  set(target, cwd, output) {
+    if (target.functionMatch) {
+      this.functions = ((target.functionMatch instanceof Array) ? target.functionMatch : [ target.functionMatch ]).filter(f => {
+        if (typeof f !== 'function') {
+          this.emit('error', 'found functionMatch that is no function: ' + typeof f);
+          return false;
+        }
+        return true;
+      });
+    }
+    this.regex = {
+      Error: this._prepareRegex(target.errorMatch),
+      Warning: this._prepareRegex(target.warningMatch)
```

Comment:
atom-build currently prints `Error` capitalized, I'm generating that name from the field name.